x-pack/plugins/enterprise_search/server/lib/enterprise_search_request_handler.ts

```diff
@@ -126,6 +130,36 @@ export class EnterpriseSearchRequestHandler {
     };
   }
 
+  /**
+   * This path helper is similar to React Router's generatePath, but much simpler &
+   * does not use regexes. It enables us to pass a static '/foo/:bar/baz' string to
+   * createRequest({ path }) and have :bar be automatically replaced by the value of
+   * request.params.bar.
+   * It also (very importantly) wraps all URL request params with encodeURIComponent(),
+   * which is an extra layer of encoding required by the Enterprise Search server in
+   * order to correctly & safely parse user-generated IDs with special characters in
+   * their names - just encodeURI alone won't work.
+   */
+  encodePathParams(path: string, params: UnknownObject) {
```

Comment:
## Why did I go with this approach?

I wrestled with a couple of different approaches for this and ended up on this one because it was both a fix and a nice enhancement / simplification for some of our existing route handlers.

### Thonk 1

One quick and dirty fix I considered was removing the `encodeURI` around `path` and just manually calling `encodeURIComponent()` within `routes/app_search/documents.ts`, e.g.:
```js
    async (context, request, response) => {
      const documentId = encodeURIComponent(request.params.documentId);

      return enterpriseSearchRequestHandler.createRequest({
        path: `/as/engines/${request.params.engineName}/documents/${documentId}`,
      })(context, request, response);
    }
```

I decided against this pretty quickly because it required trusting devs to manually remembering that certain URL params need to be encoded, it was repetitive, and it also overly trusts that `engineName` is cleaned both from the client-side and from the Enterprise Search server.

I wanted an automatic solution that encoded all dynamic params with user-set data.

### Thonk 2

Here's an example of my first naive attempt at automatic encoding:

https://github.com/constancecchen/kibana/commit/b1cfbc3f8300f566ca4341592e8237d585686410

This did work pretty OK, but I found to my dismay that IDs with slashes in them e.g. `test/this/id` worked, but weren't being properly escaped to `%2F` (because I was splitting on `/`s). This felt like a potential breaking issue or bug for APIs that that looked like `/foo/:bar/baz`.

### Thonk 3 (this one)

I sat down and thought some more and realized essentially what I wanted was something similar to React Router's [generatePath](https://reactrouter.com/core/api/generatePath) functionality. I actually legitimately tried importing and using `generatePath` but unfortunately realized that generatePath was doing some opinionated encoding of its own, which I did not want 😞 

So in the end, I essentially made my own bootleg version of generatePath. It works surprisingly well for what it is and I'm definitely pleased that it cleans up a lot of this that we were repeating over and over:

```js
    async (context, request, response) => {
      return enterpriseSearchRequestHandler.createRequest({
        path: `/engines/${request.params.engineName}/something`,
      })(context, request, response);
    }
```
which simply becomes:
```js
    enterpriseSearchRequestHandler.createRequest({
      path: '/engines/:engineName/something',
    })
```

I figured since I was here, I might as well pick an approach that cleans up / DRYs out our server API requests as well as fixes a bug. Hope y'all agree, but LMK if not!

Reply:
Also @scottybollinger Just FYI I have another PR follow-up to this that updates all of WS's routes to use this new helper/path format as well. I left it as a separate PR because I would prefer y'all regression test WS separately (and also there's a lot of separate comments/cleanup I want to make)